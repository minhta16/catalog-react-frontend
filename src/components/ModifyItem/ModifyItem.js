import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Redirect, Link } from 'react-router-dom';
import { selectCurrentUserPost, selectCurrentUserProp } from 'selectors/users';
import { selectCategories } from 'selectors/categories';
import {
  modifyPostAndRefetch as modifyPostAndRefetchRedux,
  addPostAndRefetch as addPostAndRefetchRedux,
} from 'actions/posts';
import { Typography, Container, Paper, TextField, Button, MenuItem, Grid } from '@material-ui/core';
import BackupIcon from '@material-ui/icons/Backup';
import CancelIcon from '@material-ui/icons/Cancel';

export class ModifyItem extends Component {
  state = {
    title: '',
    description: '',
    editing: false,
    redirect: false,
    // eslint-disable-next-line react/destructuring-assignment
    selectedCategory: this.props.categories.length ? this.props.categories[0].id : '0',
  };

  handleChange = (e) => {
    this.setState({
      [e.target.id]: e.target.value,
    });
  };

  handleOnSubmit = (e) => {
    e.preventDefault();

    const { token, match, modifyPostAndRefetch, addPostAndRefetch } = this.props;
    const { title, description, editing, selectedCategory } = this.state;

    if (editing) {
      modifyPostAndRefetch(token, match.params.id, match.params.postId, {
        name: title,
        description,
        price: 0,
      });
    } else {
      addPostAndRefetch(token, selectedCategory, {
        name: title,
        description,
        price: 0,
      });
    }
    this.setState({
      redirect: true,
    });
  };

  handleCategoryChange = (e) => {
    this.setState({
      [e.target.name]: e.target.value,
    });
  };

  componentDidMount = () => {
    const { match, post } = this.props;
    if (match.params.id) {
      this.setState(
        {
          editing: true,
        },
        () => {
          this.setState({
            title: post.name,
            description: post.description,
            selectedCategory: match.params.id,
          });
        },
      );
    }
  };

  render() {
    const { title, description, editing, redirect, selectedCategory } = this.state;
    const { match, categories } = this.props;

    if (redirect) {
      let path = {
        pathname: `/${match.params.id}/${match.params.postId}`,
        snackbarMess: 'Post edited!',
      };
      if (!editing) {
        path = { pathname: `/${selectedCategory}`, snackbarMess: 'Post created!' };
      }
      return <Redirect exact to={path} />;
    }
    return (
      <Container maxWidth="lg">
        <Paper className="paper">
          {editing ? (
            <Typography variant="h4">Edit Post</Typography>
          ) : (
            <Typography variant="h4">New Post</Typography>
          )}
          <form onSubmit={this.handleOnSubmit} autoComplete="off">
            <TextField
              id="categories"
              name="selectedCategory"
              select
              className="left-margin"
              label="Category"
              fullWidth
              value={selectedCategory}
              onChange={this.handleCategoryChange}
              disabled={editing}
              required
              margin="normal"
            >
              {categories.map((category) => (
                <MenuItem key={category.id} value={category.id}>
                  {category.name}
                </MenuItem>
              ))}
            </TextField>
            <TextField
              required
              id="title"
              label="Title"
              value={title}
              onChange={this.handleChange}
              margin="normal"
              fullWidth
            />
            <TextField
              required
              id="description"
              label="Description"
              value={description}
              onChange={this.handleChange}
              margin="normal"
              fullWidth
              multiline
            />
            <Grid container justify="space-evenly">
              <Button variant="contained" color="primary" type="submit">
                <BackupIcon className="icon-padding" />
                Submit
              </Button>
              <Button variant="contained" component={Link} exact="true" to="/profile">
                <CancelIcon className="icon-padding" />
                Cancel
              </Button>
            </Grid>
          </form>
        </Paper>
      </Container>
    );
  }
}

ModifyItem.propTypes = {
  match: PropTypes.object.isRequired,
  post: PropTypes.object,
  token: PropTypes.string.isRequired,
  categories: PropTypes.array.isRequired,
  addPostAndRefetch: PropTypes.func.isRequired,
  modifyPostAndRefetch: PropTypes.func.isRequired,
};

ModifyItem.defaultProps = {
  post: {},
};

const mapStateToProps = (state, ownProps) => ({
  post: selectCurrentUserPost(state, ownProps.match.params.postId),
  token: selectCurrentUserProp(state, 'token'),
  categories: selectCategories(state),
});

const mapDispatchToProps = {
  modifyPostAndRefetch: modifyPostAndRefetchRedux,
  addPostAndRefetch: addPostAndRefetchRedux,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ModifyItem);
