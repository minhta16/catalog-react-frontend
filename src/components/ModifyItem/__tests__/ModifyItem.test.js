/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import { ModifyItem } from '../ModifyItem';

describe('components/ModifyItem', () => {
  let wrapper;
  let props;
  // let warning;

  const update = () => {
    wrapper.update();
  };

  const setup = () => {
    wrapper = shallow(<ModifyItem {...props} />);
  };

  beforeEach(() => {
    props = {
      match: {
        params: {
          id: '1',
          postId: '1',
        },
      },
      post: {
        name: 'haha',
        description: 'haha',
        id: '1',
        category_id: '1',
      },
      categories: [
        {
          name: 'haha',
          description: 'haha',
          id: '1',
        },
      ],
      token: 'minh',
      modifyPostAndRefetch: jest.fn(),
      addPostAndRefetch: jest.fn(),
    };
  });

  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  describe('state.selectedCategory', () => {
    it('should be 0 when categories has no elements', () => {
      props.match.params = {};
      props.categories = [];
      setup();
      expect(wrapper.state().selectedCategory).toBe('0');
    });
  });

  it('should change state when a field is changed', () => {
    setup();
    const titleField = wrapper.find('#title');
    titleField.simulate('change', {
      target: {
        value: 'meomeo',
        id: 'title',
      },
    });
    expect(wrapper.state().title).toBe('meomeo');
  });

  it('should not have title and desc when id is not available', () => {
    props.match.params.id = '';
    setup();
    expect(wrapper.state().editing).toBe(false);
  });

  it('should call modifyPostAndRefetch() if editing', () => {
    setup();
    wrapper.setState({
      editing: true,
    });
    update();
    wrapper.instance().handleOnSubmit({ preventDefault: jest.fn() });
    expect(props.modifyPostAndRefetch).toHaveBeenCalled();
  });

  it('should call addPostAndRefetch() if not editing', () => {
    setup();
    wrapper.setState({
      editing: false,
    });
    update();
    wrapper.instance().handleOnSubmit({ preventDefault: jest.fn() });
    expect(props.addPostAndRefetch).toHaveBeenCalled();
  });

  it('should change state when handleCategoryChange() is called', () => {
    setup();
    const categoryDrop = wrapper.find('#categories');
    categoryDrop.simulate('change', {
      target: {
        name: 'selectedCategory',
        value: 99,
      },
    });
    expect(wrapper.state().selectedCategory).toBe(99);
  });
});
