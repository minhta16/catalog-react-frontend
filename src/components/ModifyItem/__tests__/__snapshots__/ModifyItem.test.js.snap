// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ModifyItem should render correctly 1`] = `
<WithStyles(ForwardRef(Container))
  maxWidth="lg"
>
  <WithStyles(ForwardRef(Paper))
    className="paper"
  >
    <WithStyles(ForwardRef(Typography))
      variant="h4"
    >
      Edit Post
    </WithStyles(ForwardRef(Typography))>
    <form
      autoComplete="off"
      onSubmit={[Function]}
    >
      <WithStyles(ForwardRef(TextField))
        className="left-margin"
        disabled={true}
        fullWidth={true}
        id="categories"
        label="Category"
        margin="normal"
        name="selectedCategory"
        onChange={[Function]}
        required={true}
        select={true}
        value="1"
      >
        <WithStyles(ForwardRef(MenuItem))
          key="1"
          value="1"
        >
          haha
        </WithStyles(ForwardRef(MenuItem))>
      </WithStyles(ForwardRef(TextField))>
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        id="title"
        label="Title"
        margin="normal"
        onChange={[Function]}
        required={true}
        value="haha"
      />
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        id="description"
        label="Description"
        margin="normal"
        multiline={true}
        onChange={[Function]}
        required={true}
        value="haha"
      />
      <WithStyles(ForwardRef(Grid))
        container={true}
        justify="space-evenly"
      >
        <WithStyles(ForwardRef(Button))
          color="primary"
          type="submit"
          variant="contained"
        >
          <BackupIcon
            className="icon-padding"
          />
          Submit
        </WithStyles(ForwardRef(Button))>
        <WithStyles(ForwardRef(Button))
          component={[Function]}
          exact="true"
          to="/profile"
          variant="contained"
        >
          <CancelIcon
            className="icon-padding"
          />
          Cancel
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(Grid))>
    </form>
  </WithStyles(ForwardRef(Paper))>
</WithStyles(ForwardRef(Container))>
`;
